src/libpanic_unwind/dwarf/eh.rs

```diff
@@ -121,15 +121,17 @@ pub unsafe fn find_eh_action(lsda: *const u8, context: &EHContext<'_>)
                 // Can never have null landing pad for sjlj -- that would have
                 // been indicated by a -1 call site index.
                 let lpad = (cs_lpad + 1) as usize;
-                return Ok(interpret_cs_action(cs_action, lpad))
+                return Ok(interpret_cs_action(cs_action, lpad, foreign_exception))
             }
         }
     }
 }
 
-fn interpret_cs_action(cs_action: u64, lpad: usize) -> EHAction {
+fn interpret_cs_action(cs_action: u64, lpad: usize, foreign_exception: bool) -> EHAction {
     if cs_action == 0 {
         EHAction::Cleanup(lpad)
+    } else if foreign_exception {
+        EHAction::None
```

Comment:
I believe C++ does catch foreign exceptions using `try { ... } catch(exception e) { ... }`.

Reply:
Yes but `catch_unwind` needs to return a `Box<Any>` when it catches an exception, which we can only do for Rust panics since those hold the `Box<Any>` as the exception payload.

Reply:
For other exceptions I think we could use some opaque `ForeignExceptionPayload` type containing the payload received from the unwinder as type behind the `Box<Any>`. And maybe even unwrap it again when calling `std::panic::resume_unwind`.

Reply:
@bjorn3 Not that this impacts your overall point, but C++ should only be able to catch foreign exceptions with `catch {` or `catch (null*) {`

Reply:
(small correction regarding the C++ syntax: it's `catch (...)` that can catch foreign exceptions)

That's unfortunately not possible for several reasons:
- The foreign exception object is completely opaque (since we don't know its real type) and the only operations we can do with it is rethrow it or free it.
- The foreign exception object has a lifetime limited to the `catch` clause that caught it. In Rust, the catch block is entirely hidden in the `intrinsics::try` compiler builtin.

In summary, we only have 2 choices: let foreign exceptions silently unwind through `catch_unwind`, or catch the foreign exception and immediately free it, then report that "some foreign exception happened" through a `Box<ForeignException>`.`ForeignException` would be a ZST since at that point it is too late to rethrow the exception (we've left the `catch` block).

Reply:
Or abort.

Reply:
I don't think this PR is the right place to resolve this question though

Reply:
C++ can catch foreign exceptions with both `catch(...) { ... }` _and_ `catch(__foreign_exception& e) { ... }` as well, where the `__foreign_exception` types can be exposed in `<exception>` header and can have sub-classes like `__java_exception` - which types are available is implementation-defined.

See, for example, [the Itanium C++ EH ABI section 1.6.4](https://itanium-cxx-abi.github.io/cxx-abi/abi-eh.html#base-personality):  

> Example: Foreign Exceptions in C++. In C++, foreign exceptions can be caught by a `catch(...)` statement. They can also be caught as if they were of a `__foreign_exception` class, defined in `<exception>`. The `__foreign_exception` may have subclasses, such as `__java_exception` and `__ada_exception`, if the runtime is capable of identifying some of the foreign languages.


>> let foreign exceptions silently unwind through catch_unwind, or catch the foreign exception and immediately free it, then report that "some foreign exception happened" through a
>
> Or abort.
>
> I don't think this PR is the right place to resolve this question though

Shall we open an issue in project-ffi-unwind to track this ? Or what would be the appropriate forum to discuss it ? I think we should try to at least properly document this behavior somewhere at some point, and that it would be worth to continue discussing which options we have for defining it somewhere.